### ISSUE TITLE
v0.13.0 doesn't resolve files loaded by require()

### ISSUE THREAD

**Issue by http91** [tststs/atom-ternjs#200]
Hi,

I've just upgraded it to the latest version and autocomplete feature for modules loaded by require() doesn't seem to work.

var async = require('async');

Angular, node and node-express works fine, they're very fast, which is great! :)

Here's my .tern-project file

```
{
  "ecmaVersion": 6,
  "libs": [],
  "dontLoad": [
    // there's a big array contains package names that I don't use. 
   // I work with ~10 packages, so I excluded the rest of them.
  ],
  "plugins": {
    "node": {},
    "node-express": {},
    "requirejs": {},
    "angular": {},
    "modules": {},
    "es_modules": {},
    "doc_comment": {
      "fullDocs": true
    }
  }
}
```

**Comment by tststs** [comment]
Please test 0.12.1 and restart all instances.
If the issue still exists after updating feel free to re-open the issue.

Greetings.

**Comment by http91** [comment]
I tried both 0.13.1 and 0.13.2 (closed Atom, restarted ternjs server after I open it) but I still don't see any autocompletions for files/modules I include using require().

**Comment by tststs** [comment]
Did completion for `async` in this specific case worked before 0.13.0?

**Comment by http91** [comment]
Yes.

I had a backup of my old atom-ternjs (0.12.2) and if I revert it to 0.12.2, it works fine. It's not just async, I can't see autocompletions for other files like below, too.

controllers/feedController.js

```
exports.index = function (req, res) {
    // functionality
};

exports.newPost = function (req, res) {
    // functionality
};

exports.editPost = function (req, res) {
    // functionality
};
```

routes/feed.js

```
var express = require('express');
var router = express.Router();
var feedController = require('../controllers/feedController');

router.get('/feed', feedController.);
```

See, if I type feedController and dot, I should get exported methods like index, newPost, editPost, etc. The same thing goes on with third party Node.js modules (not core modules) like Bookshelf.js, async, Redis, multer, etc.

**Comment by tststs** [comment]
The problem is, that the `node_resolve` plugin of tern uses require to locate the required module. Since there is no require in a webworker tern is unable to find e.g. `async`. I don't know, if this can be solved via using `require.js` in the webworker or some other solution.

For now, you can downgrade or use the relative path to the required module.
E.g.:

`var async = require('./node_modules/async/lib/async.js');`

**Comment by http91** [comment]
Got it.

It worked fine for async and entities, but -I think- since some modules have dependencies (they don't use .js files, just paths), this didn't work for bookshelf and knex. Maybe there can be a way to do this, though, but I'm unable to do at this moment.

For now, I downgraded to 0.12.2.

Thanks again.

**Comment by tststs** [comment]
There is a new branch called `node-fork`.You may try this one. I don't know if third party plugins are working.Third party plugins always are a huge pain in the ass, sometimes they work, sometime not, never managed to find out why there are working or not :)

Feedback would be nice.

Greetings.

**Comment by http91** [comment]
I downloaded zip from the repository, copied into ~/.atom/packages/atom-ternjs and installed dependencies via npm install.

I didn't want to open a new issue since it's not an official release yet. I ran into some issues :) I wish I could help more, but for now I'm limited.

[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 1.5.0-beta0
**System**: Mac OS X 10.11.3
**Thrown From**: [atom-ternjs](https://github.com/tststs/atom-ternjs) package, v0.13.2


### Stack Trace

Uncaught TypeError: this.resolves[e.id] is not a function

```
At /Users/master/.atom/packages/atom-ternjs/lib/atom-ternjs-server.js:148

TypeError: this.resolves[e.id] is not a function
    at Server.onWorkerMessage (/Users/master/.atom/packages/atom-ternjs/lib/atom-ternjs-server.js:148:26)
    at emitTwo (events.js:87:13)
    at ChildProcess.emit (events.js:172:7)
    at handleMessage (internal/child_process.js:685:10)
    at Pipe.channel.onread (internal/child_process.js:440:11)
```

### Commands

```
     -1:17.8.0 emmet:insert-formatted-line-break-only (atom-text-editor.editor.is-focused)
     -1:17.8.0 editor:newline (atom-text-editor.editor.is-focused)
     -1:17.7.0 emmet:insert-formatted-line-break-only (atom-text-editor.editor.is-focused)
     -1:17.7.0 editor:newline (atom-text-editor.editor.is-focused)
  5x -1:12.9.0 core:backspace (atom-text-editor.editor.is-focused)
     -1:08.8.0 core:move-up (atom-text-editor.editor.is-focused)
     -0:02.8.0 emmet:insert-formatted-line-break-only (atom-text-editor.editor.is-focused)
     -0:02.8.0 editor:newline (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "disabledPackages": [
      "language-csharp",
      "language-c",
      "language-clojure",
      "language-go",
      "language-java",
      "language-mustache",
      "language-make",
      "language-objective-c",
      "language-perl",
      "language-python",
      "language-ruby",
      "language-ruby-on-rails",
      "language-yaml",
      "language-toml",
      "spell-check",
      "status-bar",
      "welcome",
      "editor-stats",
      "symbols-view",
      "language-xml",
      "package-generator",
      "atom-ctags"
    ]
  },
  "atom-ternjs": {
    "excludeLowerPriorityProviders": true,
    "lint": false
  }
}
```

### Installed Packages

```coffee
# User
angularjs, v0.3.4
atom-ternjs, v0.13.2
autocomplete-angular-material, v0.4.0
autocomplete-modules, v1.3.1
color-picker, v2.0.14
emmet, v2.3.18
file-icons, v1.6.14
pretty-json, v0.4.1

# Dev
No dev packages
```

**Comment by mikeys** [comment]
Experiencing the same issues mentioned by @http91 for 0.13.2. 0.12.2 works fine.

**Comment by http91** [comment]
@tststs , @mikeys I created a page for those who want to take advantage of autocompletion and alleviate the lag issue, you may want to check it out. I think we can keep using 0.12.2 until @tststs release a new version.

https://github.com/tststs/atom-ternjs/wiki/For-those-who-want-fast-autocompletion-for-third-party-node-packages

**Comment by tststs** [comment]
I merged the node fork version into master.
Haven't had any time to test this on windows machines yet, so there will be no release until then.

Fell free to clone the repo and test it. Any feedback welcome.

1. clone repo into `.atom/packages/atom-ternjs`
2. `$ apm install`
3. restart atom

**Comment by http91** [comment]
@tststs 
I've done that (copied master repository into atom-ternjs, apm install, restarted Atom), but unfortunately I can't see any autocompletions.

**Comment by tststs** [comment]
any error in the console? you may need to uninstall atom-ternjs and then redo the steps and have a look under preferences/packages if the plugin is registered. i'm not entirely sure.

**Comment by http91** [comment]
@tststs 
I don't see any error logs.
I uninstalled & installed again by copying the repo into atom-ternjs folder and ran apm install, but nothing happens when I press something like app.ge (app.get) or app.pos (app.post) in the editor. Still playing with it, though, I will post updates here.

**Comment by tststs** [comment]
I'm using this version for quite some time now and everything works as expected for me.
are you using any third party plugins for tern (e.g. node-express)?

**Comment by http91** [comment]
Yes, here's my .tern-project file. 

I tried with and without node-express, but the result is the same. One more thing, when I view master repo, the last update is shown as a month ago (february 4). I'm not experienced with Github and repos, though, just wanted to clarify things. There's nothing missing in the current repo, right?

```
{
  "ecmaVersion": 5,
  "libs": [],
  "plugins": {
    "node": {},
    "node-express": {},
    "requirejs": {},
    "angular": {},
    "modules": {},
    "es_modules": {},
    "doc_comment": {
      "fullDocs": false
    }
  }
}

```

**Comment by tststs** [comment]
the current version might not work, if you don't specify the location of your apps JavaScript. so please configure the project via `packages -> atom-ternjs -> configure project` and restart the server.

**Comment by philiiiiiipp** [comment]
Hey, I experience the same issues. My .tern-project:

```javascript
{
  "ecmaVersion": 6,
  "libs": [],
  "loadEagerly": [
    "*.js",
    "database/*.js",
    "routes/*.js"
  ],
  "plugins": {
    "doc_comment": {
      "fullDocs": true
    }
  }
}
```

It seems as if the moment I add `module.exports = MyJsClass;` and restart the server, ternjs stops working for that file.

**Comment by http91** [comment]
1-) I download master repo, unzip, copy and paste into .atom/packages/atom-ternjs
2-) apm install. (Installing modules returns ✓ )
3-) I create .tern-project file like below.

```
{
	"ecmaVersion": 5,
	"libs": [],
	"loadEagerly": [
		"./**/*.js"
	],
	"plugins": {
		"doc_comment": {
			"fullDocs": true
		},
		"node": {},
		"requirejs": {},
		"angular": {},
		"modules": {},
		"es_modules": {}
	}
}
```

app.js test file

```
var http = require('http');
var fs = require("fs");

// fs.
// http. (nothing happens. no error, no lag..)
```

**Comment by czerwonkabartosz** [comment]
@tststs Thank you, your advice works.
@http91 Maybe you can try use my configuration and after the restart Atom, you must wait for 30 seconds to 1 minute and then everything works

> 
> I merged the node fork version into master.
> Haven't had any time to test this on windows machines yet, so there will be no release until then.
> 
> Fell free to clone the repo and test it. Any feedback welcome.
> 
> clone repo into .atom/packages/atom-ternjs
> $ apm install
> restart atom

Here is my Atom configuration - https://gist.github.com/czerwonkabartosz/4bc2df65d542b367c239

```
{
  "ecmaVersion": 5,
  "libs": [],
  "loadEagerly": [
    "lib/**/*.js"
  ],
  "plugins": {
    "doc_comment": {
      "fullDocs": true
    },
    "node": {},
    "requirejs": {},
    "angular": {},
    "modules": {},
    "es_modules": {},
    "lint": {}
  }
}

```

**Comment by petemill** [comment]
So is it now impossible for the following?

 - not specify *anything* in loadEagerly
 - node plugin
 - `require('./mymodule')` (no '.js' ending)

It seems that the only problem is that the plugin is not looking on the filesystem to see if it can find `path`.js or `path`/index.js when it encounters a module. Having that functionality would be amazing.

**Comment by brunolemos** [comment]
Time for a new release? :)

**Comment by philiiiiiipp** [comment]
Hey, I think this issue is fixed? (Tested with 0.14.1)

**Comment by http91** [comment]
Yes, it is fixed. I tested and it works great. I've been waiting for this. It's time to ditch WebStorm for me, again.

Hello, Atom.
